src/arithmetic/fields.rs

```diff
@@ -103,6 +140,201 @@ pub trait FieldExt:
     }
 }
 
+/// Parameters for a perfect hash function used in square root computation.
+#[derive(Debug)]
+struct SqrtHasher<F: FieldExt> {
+    hash_xor: u32,
+    hash_mod: usize,
+    marker: PhantomData<F>,
+}
+
+impl<F: FieldExt> SqrtHasher<F> {
+    /// Returns a perfect hash of x for use with SqrtTables::inv.
+    fn hash(&self, x: &F) -> usize {
+        // This is just the simplest constant-time perfect hash construction that could
+        // possibly work. The 32 low-order bits are unique within the 2^S order subgroup,
+        // then the xor acts as "salt" to injectively randomize the output when taken modulo
+        // `hash_mod`. Since the table is small, we do not need anything more complicated.
+        ((x.get_lower_32() ^ self.hash_xor) as usize) % self.hash_mod
+    }
+}
+
+/// Tables used for square root computation.
+#[derive(Debug)]
+pub struct SqrtTables<F: FieldExt> {
+    hasher: SqrtHasher<F>,
+    inv: Vec<u8>,
+    g0: Box<[F; 256]>,
+    g1: Box<[F; 256]>,
+    g2: Box<[F; 256]>,
+    g3: Box<[F; 129]>,
+}
+
+impl<F: FieldExt> SqrtTables<F> {
+    /// Build tables given parameters for the perfect hash.
+    pub fn new(hash_xor: u32, hash_mod: usize) -> Self {
+        let hasher = SqrtHasher {
+            hash_xor,
+            hash_mod,
+            marker: PhantomData,
+        };
+
+        let gtab: Vec<Vec<F>> = (0..4)
+            .scan(F::ROOT_OF_UNITY, |gi, _| {
+                // gi == ROOT_OF_UNITY^(256^i)
+                let gtab_i: Vec<F> = (0..256)
+                    .scan(F::one(), |acc, _| {
+                        let res = *acc;
+                        *acc *= *gi;
+                        Some(res)
+                    })
+                    .collect();
+                *gi = gtab_i[255] * *gi;
+                Some(gtab_i)
+            })
+            .collect();
+
+        // Now invert gtab[3].
+        let mut inv: Vec<u8> = vec![1; hash_mod];
+        for j in 0..256 {
+            let hash = hasher.hash(&gtab[3][j]);
+            // 1 is the last value to be assigned, so this ensures there are no collisions.
+            assert!(inv[hash] == 1);
+            inv[hash] = ((256 - j) & 0xFF) as u8;
+        }
+
+        SqrtTables::<F> {
+            hasher,
+            inv,
+            g0: Box::new(gtab[0][..].try_into().unwrap()),
+            g1: Box::new(gtab[1][..].try_into().unwrap()),
+            g2: Box::new(gtab[2][..].try_into().unwrap()),
+            g3: Box::new(gtab[3][0..129].try_into().unwrap()),
```

Comment:
This will temporarily occupy a large portion of the stack before Rust actually copies it into the heap, which is annoying but can be mitigated by doing e.g. `gtab[0].into_boxed_slice()` though you will lose static slice lengths without using `unsafe` to restore them.

I suspect that when we bring the pasta impl out into a separate crate eventually we will need to feature flag this entire sqrt implementation so that users who are trying to compile this for embedded devices can turn off these massive precomputed tables. So, for now, let's leave this code alone.

Reply:
I considered using `into_boxed_slice`, but it's not worth the unsafety.

I'd rather just reduce the table size than switch off this implementation for embedded devices. With 16-element tables it's still much more efficient than Tonelli-Shanks.